File: Neato-SDK/neato-sdk-android/src/main/java/com/neatorobotics/sdk/android/authentication/NeatoAuthentication.java
package com.neatorobotics.sdk.android.authentication;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;

import com.neatorobotics.sdk.android.R;

import java.util.Date;

/**
 * Neato-SDK
 * Created by Marco on 06/05/16.
 * Copyright © 2016 Neato Robotics. All rights reserved.
 */
public class NeatoAuthentication {

    private static NeatoAuthentication instance = null;
    private AccessTokenDatasource accessTokenDatasource;

    /**
     * Use this method to get the singleton that use the default access token datasource.
     * @param context
     * @return
     */
    public static NeatoAuthentication getInstance(Context context) {
        if(instance == null) {
            instance = new NeatoAuthentication(context);
        }
        return instance;
    }

    /**
     * Use this method to get the singleton that use a custom access token datasource.
     * @param context
     * @param accessTokenDatasource
     * @return
     */
    public static NeatoAuthentication getInstance(Context context, AccessTokenDatasource accessTokenDatasource) {
        NeatoAuthentication auth = getInstance(context);
        auth.accessTokenDatasource = accessTokenDatasource;
        return auth;
    }

    /**
     *
     * @param context
     */
    private NeatoAuthentication(Context context){
        this.accessTokenDatasource = new DefaultAccessTokenDatasource(context);
    }

    /**
     * Start the OAuth 2 authentication flow
     */
    public void openLoginInBrowser(Context context,String clientId, String redirectUri, NeatoOAuth2Scope[] scopes) {
        String authUrl = buildOAuthAuthenticationUrl(context.getString(R.string.oauth2_authentication_url),clientId,buildScopesParameter(scopes), redirectUri);
        Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(authUrl));
        context.startActivity(intent);

        clearAccessToken();
    }

    /**
     * Parse the OAuth2 response starting from the response URI
     * @param uri
     * @return the NeatoAuthenticationResponse type
     */
    public NeatoAuthenticationResponse getOAuth2AuthResponseFromUri(Uri uri) {
        NeatoAuthenticationResponse response = NeatoAuthenticationResponse.fromUri(uri);
        switch (response.getType()) {
            // Response was successful and contains auth token
            case TOKEN:
                setOauth2AccessToken(response.getToken(), response.getTokenExpirationDate());
                break;
            case ERROR:
                clearAccessToken();
                break;
            // Most likely auth flow was cancelled
            default:
                //You can do nothing
        }
        return response;
    }

    /**
     *
     * @return the OAuth 2 authentication string URL
     */
    protected String buildOAuthAuthenticationUrl(String formattedUrl, String clientId, String scopes, String redirectUri) {
        String authUrl = String.format(formattedUrl, clientId, scopes, redirectUri);
        return authUrl;
    }

    /**
     *
     * @param scopes
     * @return the comma separated string of scopes
     */
    protected String buildScopesParameter(NeatoOAuth2Scope[] scopes) {
        StringBuffer scopesBuffer = new StringBuffer("");
        if(scopes != null && scopes.length > 0) {
            for (NeatoOAuth2Scope scope : scopes) {
                scopesBuffer.append(scope);
                scopesBuffer.append(",");
            }
            scopesBuffer.deleteCharAt(scopesBuffer.length()-1);
        }
        return scopesBuffer.toString();
    }

    /**
     * Use this method to check if the current user is authenticated
     * @return true if the client has a stored access token
     */
    public boolean isAuthenticated() {
        return accessTokenDatasource.isTokenValid();
    }

    /**
     * Clear the current OAuth2 access token
     */
    public void clearAccessToken(){
        accessTokenDatasource.clearToken();
    }

    /**
     * Call this method after the end of a successful OAuth2.0 authentication flow. This method setup
     * the NeatoClient with the obtained access token.
     * @param token
     * @param tokenExpirationDate
     */
    public void setOauth2AccessToken(String token, Date tokenExpirationDate) {
        accessTokenDatasource.storeToken(token, tokenExpirationDate);
    }

    /**
     * Use this method to get the current OAuth2 access token
     * @return
     */
    public String getOauth2AccessToken() {
        return accessTokenDatasource.loadToken();
    }
}

File: Neato-SDK/neato-sdk-android/src/main/java/com/neatorobotics/sdk/android/authentication/NeatoAuthentication.kt
/*
 * Copyright (c) 2019.
 * Neato Robotics Inc.
 */

package com.neatorobotics.sdk.android.authentication

import android.content.Context
import android.content.Intent
import android.net.Uri
import java.io.InvalidObjectException

import java.util.Date

object NeatoAuthentication {

    private val OAUTH_AUTH_URL = "https://apps.neatorobotics.com/oauth2/authorize?client_id=%1\$s&amp;scope=%2\$s&amp;response_type=token&amp;redirect_uri=%3\$s"

    var accessTokenDatasource: AccessTokenDatasource? = null

    /**
     * Use this method to check if the current user is authenticated
     * @return true if the client has a stored access token
     */
    val isAuthenticated: Boolean
        get() = accessTokenDatasource!!.isTokenValid

    /**
     * Use this method to get the current OAuth2 access token
     * @return
     */
    val oauth2AccessToken: String?
        get() = accessTokenDatasource!!.loadToken()

    init {
        this.accessTokenDatasource = DefaultAccessTokenDatasource()
    }

    /**
     * Start the OAuth 2 authentication flow
     */
    fun openLoginInBrowser(context: Context, clientId: String, redirectUri: String, scopes: Array<NeatoOAuth2Scope>) {
        val authUrl = buildOAuthAuthenticationUrl(OAUTH_AUTH_URL, clientId, buildScopesParameter(scopes), redirectUri)
        val intent = Intent(Intent.ACTION_VIEW, Uri.parse(authUrl))
        context.startActivity(intent)

        clearAccessToken()
    }

    /**
     * Parse the OAuth2 response starting from the response URI
     * @param uri
     * @return the NeatoAuthenticationResponse type
     */
    fun getOAuth2AuthResponseFromUri(uri: Uri): NeatoAuthenticationResponse {
        val response = NeatoAuthenticationResponse.fromUri(uri)
        when (response.type) {
            // Response was successful and contains auth token
            NeatoAuthenticationResponse.Response.TOKEN -> setOauth2AccessToken(
                response.token?:"",
                response.tokenExpirationDate?:Date()
            )
            NeatoAuthenticationResponse.Response.ERROR -> clearAccessToken()
        }// Most likely auth flow was cancelled
        //You can do nothing
        return response
    }

    /**
     *
     * @return the OAuth 2 authentication string URL
     */
    fun buildOAuthAuthenticationUrl(
        formattedUrl: String,
        clientId: String?,
        scopes: String,
        redirectUri: String?
    ): String {
        return String.format(formattedUrl, clientId, scopes, redirectUri)
    }

    /**
     *
     * @param scopes
     * @return the comma separated string of scopes
     */
    fun buildScopesParameter(scopes: Array<NeatoOAuth2Scope>?): String {
        val scopesBuffer = StringBuffer("")
        if (scopes != null && scopes.isNotEmpty()) {
            for (scope in scopes) {
                scopesBuffer.append(scope)
                scopesBuffer.append("+")
            }
            scopesBuffer.deleteCharAt(scopesBuffer.length - 1)
        }
        return scopesBuffer.toString()
    }

    /**
     * Clear the current OAuth2 access token
     */
    fun clearAccessToken() {
        accessTokenDatasource!!.clearToken()
    }

    /**
     * Call this method after the end of a successful OAuth2.0 authentication flow. This method setup
     * the NeatoClient with the obtained access token.
     * @param token
     * @param tokenExpirationDate
     */
    fun setOauth2AccessToken(token: String, tokenExpirationDate: Date) {
        accessTokenDatasource?.storeToken(token, tokenExpirationDate)
    }
}
